### PR TITLE
[FIXED JENKINS-28632] Workflow step for awaiting for a deployment

### PR DESCRIPTION
**Pull request by fbelzunc**
This is a Workflow step for awaiting for a deployment

https://issues.jenkins-ci.org/browse/JENKINS-28632

@reviewbybees @jglick

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/deployment/Condition.java [hunk 1]
```diff
@@ -16,8 +16,12 @@
      * of the given {@link Job}. If so, return the build number of the job that matches,
      * otherwise return negative number to indicate no match.
      */
+    @Deprecated
     public abstract RangeSet calcMatchingBuildNumberOf(Job upstream, DeploymentFacet<?> facet);
 
+    public abstract RangeSet calcMatchingBuildNumberOf(Job upstream, DeploymentFacet<?> facet, Job workflowJob);
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/deployment/Condition.java]: Backward incompatibility (this is going to break any implementation). You should implement something [like this](https://github.com/jenkinsci/analysis-core-plugin/blob/9f8f492d2931981f91a06a8eb0123fd7d3cd1a58/src/main/java/hudson/plugins/analysis/core/HealthAwarePublisher.java#L129-L145).

File: src/main/java/org/jenkinsci/plugins/deployment/conditions/ThresholdCondition.java [hunk 2]
```diff
@@ -36,6 +36,38 @@ public int getThreshold() {
         return threshold;
     }
 
+
+    @Override
+    public RangeSet calcMatchingBuildNumberOf(Job upstream, DeploymentFacet<?> facet, Job workflowJob) {
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/deployment/conditions/ThresholdCondition.java]: Why do you need two `Job` parameters? It does not matter if it's a workflow job or not.

File: src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/AwaitDeploymentStepExecution.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    @Inject(optional=true) private transient AwaitDeploymentStep awaitDeploymentStep;
+    @StepContextParameter
+    private transient FlowNode node;
+    @StepContextParameter
+    transient Run run;
+    private transient volatile ScheduledFuture<?> task;
+    @StepContextParameter
+    private transient Run build;
+    @StepContextParameter
+    private transient Job job;
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/AwaitDeploymentStepExecution.java]: All the above block is really hard to read. Both inline and top-line annotations, no new-lines between definitions, ...

File: src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/AwaitDeploymentStepExecution.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    public void proceed(DeploymentFacet<?> facet, HostRecord newRecord) {
+        ThresholdCondition thresholdCondition = new ThresholdCondition(awaitDeploymentStep.getEnv(), awaitDeploymentStep.getThreshold());
+        Fingerprint.RangeSet r = thresholdCondition.calcMatchingBuildNumberOf(null, facet.getFingerprint(), job);
+
+        if (!r.isEmpty()) {
+            for (Integer n : r.listNumbers()) {
+                Run b = job.getBuildByNumber(n);
+                if (b!=null) {
+                    if (b.getId().equals(build.getId())) {
+                        getContext().onSuccess(null);
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/AwaitDeploymentStepExecution.java]: Is it granted that this is going to be called only once? IOW, is the map `facet, newRecord` -> `build` always defined and 1:1.

File: src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/WorkflowListenerImpl.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                }
+            });
+            }
+        });
+    }
+
+    /**
+     * Waits until all the pending deployment facets are processed.
+     */
+    public void sync() throws InterruptedException {
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/WorkflowListenerImpl.java]: Where is this called?

**Comment by amuniz** [comment]
As I said in the JIRA issue, I would have implemented this as a block-scoped step.

File: src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/AwaitDeploymentStepExecution.java (hunk 4 above)
Comment (fbelzunc) [inline src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/AwaitDeploymentStepExecution.java]: Yes, I think so.

**Comment by stephenc** [comment]
:bee: looks ok to me, but somebody more familiar with workflow step writing should review

**Comment by fbelzunc** [comment]
@jglick @amuniz Can I have your final review here, please?

File: src/main/java/org/jenkinsci/plugins/deployment/Condition.java [hunk 6]
```diff
@@ -18,6 +18,7 @@
      */
     public abstract RangeSet calcMatchingBuildNumberOf(Job upstream, DeploymentFacet<?> facet);
 
+    @Deprecated
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/deployment/Condition.java]: Add a javadoc with `@deprecated` pointing to the alternative to use.

File: src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/AwaitDeploymentStepExecution.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+    @Override public void onResume() {
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/deployment/workflowsteps/AwaitDeploymentStepExecution.java]: Is this override needed?

**Comment by amuniz** [comment]
:bee:

**Comment by fbelzunc** [comment]
@reviewbybees done

**Pull request merged**